File: csm/transcript.py
import json

def _render(content):
    text_parts, tool_parts = [], []
    if isinstance(content, str):
        return content, []
    if isinstance(content, list):
        for b in content:
            if not isinstance(b, dict):
                continue
            bt = b.get("type")
            if bt == "text" and b.get("text"):
                text_parts.append(b["text"])
            elif bt == "tool_use":
                tool_parts.append({"name": b.get("name", "tool"),
                                   "input": b.get("input", {})})
            elif bt == "tool_result":
                tool_parts.append({"name": "result", "input": b.get("content", "")})
    return " ".join(text_parts), tool_parts

def load_transcript(file_path, limit=None):
    msgs = []
    with open(file_path, encoding="utf-8", errors="replace") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                o = json.loads(line)
            except Exception:
                continue
            if o.get("type") not in ("user", "assistant"):
                continue
            msg = o.get("message", {})
            text, tools = _render(msg.get("content"))
            msgs.append({
                "role": o.get("type"),
                "text": text,
                "tools": tools,
                "has_tool": bool(tools),
                "timestamp": o.get("timestamp", ""),
            })
    if limit:
        msgs = msgs[-limit:]
    return msgs

File: lib/transcript.js
const fs = require("fs");

const SYS = ["<task-notification>", "Base directory for this skill", "Caveat:",
  "<system-reminder>", "<command-", "<local-command", "<user-",
  "[Request interrupted", "Result of calling", "<bash-", "API Error"];

function isHuman(t) {
  t = (t || "").trim();
  if (!t) return false;
  return !SYS.some((s) => t.startsWith(s));
}

function render(content) {
  const text = [];
  const tools = [];
  let isToolResult = false;
  if (typeof content === "string") return { text: content, tools, isToolResult };
  if (Array.isArray(content)) {
    for (const b of content) {
      if (!b || typeof b !== "object") continue;
      if (b.type === "text" && b.text) text.push(b.text);
      else if (b.type === "tool_use") tools.push({ name: b.name || "tool" });
      else if (b.type === "tool_result") { tools.push({ name: "result" }); isToolResult = true; }
    }
  }
  return { text: text.join(" "), tools, isToolResult };
}

function loadTranscript(file) {
  const msgs = [];
  let data;
  try { data = fs.readFileSync(file, "utf-8"); } catch (e) { return msgs; }
  for (const line of data.split("\n")) {
    const s = line.trim();
    if (!s) continue;
    let o;
    try { o = JSON.parse(s); } catch (e) { continue; }
    if (o.type !== "user" && o.type !== "assistant") continue;
    const r = render((o.message || {}).content);
    const human = o.type === "user" && !r.isToolResult && isHuman(r.text);
    msgs.push({ role: o.type, text: r.text, has_tool: r.tools.length > 0,
      human, timestamp: o.timestamp || "" });
  }
  return msgs;
}

module.exports = { loadTranscript, isHuman };
